broker-core/src/main/java/io/zeebe/broker/logstreams/restore/SbeLogReplicationResponse.java

```diff
@@ -24,115 +24,111 @@
-
-  public SbeLogReplicationResponse(LogReplicationResponse other) {
+  private final LogReplicationResponseEncoder encoder;
+  private final LogReplicationResponseDecoder decoder;
+  private final DefaultLogReplicationResponse delegate;
+
+  public SbeLogReplicationResponse() {
+    this.delegate = new DefaultLogReplicationResponse();
+    this.encoder = new LogReplicationResponseEncoder();
+    this.decoder = new LogReplicationResponseDecoder();
     reset();
-    wrap(other);
   }
 
-  public SbeLogReplicationResponse(
-      long toPosition, boolean moreAvailable, byte[] serializedEvents) {
-    this.toPosition = toPosition;
-    this.moreAvailable = moreAvailable;
-    this.serializedEvents.wrap(serializedEvents);
+  public SbeLogReplicationResponse(LogReplicationResponse other) {
+    this();
+    delegate.setToPosition(other.getToPosition());
+    delegate.setSerializedEvents(other.getSerializedEvents());
+    delegate.setMoreAvailable(other.hasMoreAvailable());
   }
 
   public SbeLogReplicationResponse(byte[] serialized) {
-    reset();
+    this();
     wrap(new UnsafeBuffer(serialized));
   }
 
   @Override
   public void reset() {
     super.reset();
-    toPosition = toPositionNullValue();
-    moreAvailable = false;
-    serializedEvents.wrap(0, 0);
+    delegate.setToPosition(toPositionNullValue());
+    delegate.setMoreAvailable(false);
+    delegate.setSerializedEvents(new UnsafeBuffer(), 0, 0);
   }
 
   @Override
   public void wrap(DirectBuffer buffer, int offset, int length) {
     super.wrap(buffer, offset, length);
+    final DirectBuffer wrapBuffer = new UnsafeBuffer();
+    decoder.wrapSerializedEvents(wrapBuffer);
 
-    toPosition = decoder.toPosition();
-    moreAvailable = decoder.moreAvailable() == BooleanType.TRUE;
-    decoder.wrapSerializedEvents(serializedEvents);
-  }
-
-  public void wrap(LogReplicationResponse other) {
-    this.setToPosition(other.getToPosition());
-    this.setSerializedEvents(other.getSerializedEvents());
-    this.setMoreAvailable(other.hasMoreAvailable());
+    delegate.setToPosition(decoder.toPosition());
+    delegate.setMoreAvailable(decoder.moreAvailable() == BooleanType.TRUE);
+    delegate.setSerializedEvents(wrapBuffer, 0, wrapBuffer.capacity());
   }
 
   @Override
   public void write(MutableDirectBuffer buffer, int offset) {
     super.write(buffer, offset);
-    encoder.toPosition(toPosition);
-    encoder.putSerializedEvents(serializedEvents, 0, serializedEvents.capacity());
-    encoder.moreAvailable(moreAvailable ? BooleanType.TRUE : BooleanType.FALSE);
-  }
+    final byte[] serializedEvents = delegate.getSerializedEvents();
 
-  @Override
-  public int getLength() {
-    return super.getLength() + serializedEventsHeaderLength() + serializedEvents.capacity();
+    encoder.toPosition(delegate.getToPosition());
+    encoder.moreAvailable(delegate.hasMoreAvailable() ? BooleanType.TRUE : BooleanType.FALSE);
+
+    if (getSerializedEventsLength() > 0) {
```

Comment:
Question: Can it decode correctly if length is 0?

Reply:
Yes, but I guess that relies on how SBE does its encoding, so it might be safer to explicitly write a 0-length buffer anyway.